### PR TITLE
#11498: Fixed unable to saved serialized object into PostgreSQL binary column

### PR DESCRIPTION
**Pull request by klimov-paul**
| Q | A |
| --- | --- |
| Is bugfix? | yes |
| New feature? | no |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #11498, #6173 |

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -12,6 +12,7 @@ Yii Framework 2 Change Log
 - Enh #11432: Added HTTP status 421 "Misdirected Request" to list of statuses in `yii\web\Response` (dasmfm)
 - Enh #11438: Configurable `yii\helpers\Markdown` default flavor (mdmunir)
 - Bug #11459: Fixed flash messages not destroyed when `session.auto_start = 1` set in php.ini (cartmanchen)
+- Bug #11498: Fixed unable to saved serialized object into PostgreSQL binary column (klimov-paul)
```
Comment (samdark) [inline framework/CHANGELOG.md]: Fixed inability to save

File: framework/db/pgsql/QueryBuilder.php [hunk 2]
```diff
@@ -227,6 +227,41 @@ public function alterColumn($table, $column, $type)
...
+     * @param string $table the table that data will be saved into.
+     * @param array $columns the column data (name => value) to be saved into the table.
+     * @return array normalized columns
+     */
+    private function normalizeTableRowData($table, $columns)
+    {
+        if (($tableSchema = $this->db->getSchema()->getTableSchema($table)) !== null) {
+            $columnSchemas = $tableSchema->columns;
+            foreach ($columns as $name => $value) {
+                if (isset($columnSchemas[$name]) && $columnSchemas[$name]->type === Schema::TYPE_BINARY && is_string($value)) {
```
Comment (samdark) [inline framework/db/pgsql/QueryBuilder.php]: How about saving `int`s and `float`s?

Comment (klimov-paul) [inline framework/db/pgsql/QueryBuilder.php]: Currently, we have no problems with them, have we?

File: framework/CHANGELOG.md (hunk 1 above)
Comment (klimov-paul) [inline framework/CHANGELOG.md]: fixed

File: framework/db/pgsql/QueryBuilder.php (hunk 2 above)
Comment (samdark) [inline framework/db/pgsql/QueryBuilder.php]: You're right. That seems to be covered with tests.

**Comment by samdark** [comment]
Looks fine to me overall.

File: framework/db/pgsql/QueryBuilder.php [hunk 3]
```diff
@@ -227,6 +227,41 @@ public function alterColumn($table, $column, $type)
...
+    public function update($table, $columns, $condition, &$params)
+    {
+        return parent::update($table, $this->normalizeTableRowData($table, $columns), $condition, $params);
+    }
+
+    /**
+     * Normalizes data to be saved into the table, performing extra preparations and type converting, if necessary.
+     * @param string $table the table that data will be saved into.
+     * @param array $columns the column data (name => value) to be saved into the table.
+     * @return array normalized columns
```
Comment (SilverFire) [inline framework/db/pgsql/QueryBuilder.php]: @since 2.0.9

**Comment by SilverFire** [comment]
This PR also fixed https://github.com/yiisoft/yii2/issues/6173 that I've tried to fix by https://github.com/yiisoft/yii2/commit/c478199937e0c8267477da0acd60a0fb9067a13d

**Comment by SilverFire** [comment]
@klimov-paul this approach looks much better than mine

**Comment by jorge-epc** [comment]
Hi, when is going to be available 2.0.9? I've been waiting for this fix. Thanks.

**Comment by samdark** [comment]
2.0.9 will be out after http-client release.

**Pull request merged**